test/cpp/jit/tests.h

```diff
@@ -1120,6 +1120,43 @@ void testCustomOperators() {
     tracer::abandon();
   }
 }
+
+// test a few features that are not directly used in schemas yet
+void testSchemaParser() {
+  // nested arrays
+  auto s = parseSchema("at::what(int[][4] foo) -> ()");
+  ASSERT_TRUE(s.arguments.at(0).N == 4);
+  ASSERT_TRUE(IntType::get()->isSubtypeOf(s.arguments.at(0)
+                                              .type->expect<ListType>()
+                                              ->getElementType()
+                                              ->expect<ListType>()
+                                              ->getElementType()));
+  auto s2 = parseSchema("at::what(int[][] foo) -> ()");
+  ASSERT_TRUE(IntType::get()->isSubtypeOf(s2.arguments.at(0)
+                                            .type->expect<ListType>()
+                                            ->getElementType()
+                                            ->expect<ListType>()
+                                            ->getElementType()));
+  // futures
+  try {
+    parseSchema("at::what(Future(int) foo) -> ()");
+    ASSERT_TRUE(false);
+  } catch (script::ErrorReport& er) {
+    ASSERT_TRUE(
+        std::string(er.what()).find("Futures are not yet implemented") !=
+        std::string::npos);
+  }
+  // named returns
+  parseSchema("at::what(Tensor! i_will_be_written_to) -> ()");
```

Comment:
what does exclamantion mark mean?

Reply:
It means that the tensor is mutable and will be written to.

Reply:
Can we document these somewhere?